core/common/src/main/java/alluxio/underfs/UnderFileSystem.java

```diff
@@ -88,6 +91,80 @@ public int getValue() {
   }
 
   /**
+   * A class used to cache UnderFileSystems.
+   */
+  private static class Cache {
+    private ConcurrentHashMap<Key, UnderFileSystem> mUnderFileSystemMap = new ConcurrentHashMap<>();
+
+    Cache() {}
+
+    /**
+     * Gets a UFS instance from the cache if exists. Otherwise, creates a new instance and adds
+     * that to the cache.
+     *
+     * @param path the UFS path
+     * @param ufsConf the UFS configuration
+     * @param configuration the Alluxio configuration
+     * @return the UFS instance
+     */
+    UnderFileSystem get(String path, Object ufsConf, Configuration configuration) {
+      UnderFileSystem cachedFs = null;
+      Key key = new Key(new AlluxioURI(path));
+      cachedFs = mUnderFileSystemMap.get(key);
+      if (cachedFs != null) {
+        return cachedFs;
+      }
+      UnderFileSystem fs = UnderFileSystemRegistry.create(path, configuration, ufsConf);
+      cachedFs = mUnderFileSystemMap.putIfAbsent(key, fs);
+      if (cachedFs == null) {
+        return fs;
+      }
+      try {
+        fs.close();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      return cachedFs;
+    }
+  }
+
+  /**
+   * The key of the UFS cache.
+   */
+  private static class Key {
+    private final String mScheme;
+    private final String mAuthority;
+    // TODO(peis): Add UGI information.
+
+    Key(AlluxioURI uri) {
+      mScheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase();
+      mAuthority = uri.getAuthority() == null ? "" : uri.getAuthority().toLowerCase();
+    }
+
+    @Override
+    public int hashCode() {
+      return mScheme.hashCode() ^ mAuthority.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object object) {
```

Comment:
We usually follow this style for equals, then use `Objects.hashCode` to generate the hash code

```java
  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof InStreamOptions)) {
      return false;
    }
    InStreamOptions that = (InStreamOptions) o;
    return Objects.equal(mLocationPolicy, that.mLocationPolicy)
        && Objects.equal(mReadType, that.mReadType)
        && Objects.equal(mCachePartiallyReadBlock, that.mCachePartiallyReadBlock)
        && Objects.equal(mSeekBufferSizeBytes, that.mSeekBufferSizeBytes);
  }
```